### ISSUE TITLE
Push for markdown

### ISSUE THREAD

**Issue by starbelly** [erlef/build-and-packaging-wg#60]
## Background

Discuss pros and cons of making a push for markdown or asciidoc in erlang instead of edoc. 

## Objectives

- Discuss the value in moving away from edoc to a more "modern" and doc format that lowers the barrier to entry in regards to developer experience and development around said format(s) (i.e., generation of docs, output in the erl and iex, etc.) 

## Notes 

- Related EEP: https://www.erlang.org/eeps/eep-0059
- This agenda item was discussed at the last B&P meeting, this has been created post-meeting so that notes can be kept for historical record and actionables can be tracked. 
- OTP informally had agreed upon asciidoc already, however it may be worth offering markdown or even cjdot as an alternative.
     - [asciidoc](https://asciidoctor.org/) has clear pros, one big one is support for custom attributes, support for multiple backend generators (ebook, PDF, html page, etc.). It also arguably more well defined vs markdown, which some have referred to as a moving target, especially github flavored markdown. 
     - markdown is quite nice as it well established, ubiquitous, and the erlang and elixir community would be more consistent if markdown was preferred, at least at this point in time. 
     - [djot](https://djot.net/) was offered as another alternative, which has nice properties, the main one raised is that it feels more like markdown, but provides many of the features of asciidoc. Though, concerns were raised about it not being as well established as asciidoc.
- The idea of experimenting with a rebar plugin vs pushing for OTP to start with was raised vs going for buy-in via an EEP to start with was raised as well. The pros of this are fairly clear, but to note we could start adding value to the community right away vs possibly waiting quite a while for a decision on an EEP to be made. 
- An async poll was conducted, Fred Hebert, José Valim, John Högberg, Adam Lindberg, and Bryan Paxton all voted in favor of going for an EEP to get buy-in before an implementation is done. 
    - Two other options were given, however there were no other votes : 
            - Start with an experimental project that implements the work ahead of the EEP 
            - Do both at once (one person handles eep, another person gets going on the implementation)
            
 ## Edit
 
This issue started with an aim to get consensus on a proposed format. That consensus has been reached, thus the title of this issue has been updated.            
            
##  Action items

- [ ] Create and submit an EEP proposing the alternative doc format(s), how this would be implemented, and how we would execute on it all (roughly). 
- [ ] Find a volunteer for implementing the EEP should it be accepted.
- [ ] Work with the volunteer to get a stipend created defining the scope of work and work towards approval by the EEF. 
- [ ] Implement the alternative per the EEP (dependent upon it being accepted ofc). 
- [ ] Submit the work to Erlang/OTP
- [ ] Submit patch to ExDoc to support asciidoc

**Comment by michalmuskala** [comment]
As an outsider to these proceedings I have to say I'm extremely surprised anything other than markdown is even being considered. It's ubiquitous and universal - if you're an engineer you have to know markdown. This is not true of the other formats proposed.

In my view, the primary goal of the documentation syntax is to make it easy to write. I'm afraid that if Erlang goes with a less popular format, it will be the same as edoc - nobody will write documentation since they don't know the syntax. There are countless Erlang projects that have comments that could be documentation, but aren't, usually exactly because of the unfamiliarity with the syntax of edoc.

Elixir is handling documentation pretty well, if you asked many, perhaps even best-in-class, and this is without any custom attributes or metadata. Is it really necessary for Erlang? I believe many of those niche formats are an example of perfect being the enemy of the good.

**Comment by eproxus** [comment]
Markdown doesn't support attributes, which was something the OTP team asked for as far as I know. AsciiDoc and Djot does (In my opinion, Djot seems to be the best of both worlds).

**Comment by KennethL** [comment]
I have said a long time ago that AsciiDoc seems better than Markdown since it is more well defined and support meta data as I remember it. But that does not imply that Markdown can't be chosen. Maybe Markdown is the best choice for other reasons even if it has many shortcomings. It is well known, there are tools etc.

**Comment by josevalim** [comment]
FWIW, metadata can be supported externally:

    -doc #{"author" => "José Valim"}.

The advantage is that we are not tied to strings.

**Comment by lpil** [comment]
Using regular Erlang as the format for metadata sounds much more approachable than an additional syntax for folks to learn. We'd also get syntax highlighting and error messages for syntax errors, which is nice.

**Comment by eproxus** [comment]
The problem is how to replace existing inline constructs like EDoc’s `{@type foobar()}`. Markdown doesn’t support arbitrary attributes.

**Comment by lpil** [comment]
Yes that's right, José's suggestion was to use Erlang attributes for metadata.

**Comment by josevalim** [comment]
In any case, I would say even `@type` as a string without further validation is not ideal. New annotations would do a better job of validating and exposing them.

**Comment by eproxus** [comment]
Here are the ones I used in order of frequency:
1. `@equiv` (function level meta data)

**Comment by erszcz** [comment]
@eproxus
> The problem is how to replace existing inline constructs like EDoc’s {@type foobar()}. Markdown doesn’t support arbitrary attributes.

`@type` and `@spec` should be dropped altogether, since they're not supported by Dialyzer/Gradualizer. The other ones you mention: `@equiv`, `@see`, `@link`, inline `@type` are useful, but should be relatively easily replaceable by link / xrefs (depending on the format chosen). I'd add `@deprecated` and `@since` to the list, which are useful, but the support in EDoc is limited even now.

**Comment by josevalim** [comment]
Erlang could also use `see_also` and `equiv` as metadata keys. A quick update to EEP 48 would make them official too.

**Comment by DianaOlympos** [comment]
I will plug Djot in again on the list of possible choices we should consider. It has a syntax that is really close to markdown, to the point it will probably look mostly the same for most people and the differences could be limited to "differences in markdown implementation".

But it also has more support for attributes and all kind of things like these attributes we want to support.

On top of this, it is made to be easier to parse than markdown, which would make maintenance of the implementation easier. The person behind it is the author and maintainer of Pandoc and CommonMark, so he has the experience and knowledge about markdown to support the use cases.

I would totally understand if we end up not going with it, but I think it is an easier contender than Asciidoc at least, and would allow to go forward with a less painful implementation while keeping us fairly close to the objective of accessibility

**Comment by eproxus** [comment]
Would be cool if `equiv` could even be real Erlang code (maybe not possible?):
```erl
-doc #{equiv => label(T, 0)}.
label(T) ->
    label(T, 0).
```

**Comment by michalmuskala** [comment]
Elixir's ExDoc built on top of plain markdown allows references to all kinds of things - `local_function/1`, `Remote.function/2`, `t:local_type/3`, `t:Remote.type/4`, `c:local_callback/5`, `c:Remote.callback/6`. So I don't think this is a challenge.

The aspect that ASCIIDoc and Djot don't have an existing Erlang or Elixir implementation and thus impose additional toolchain complexity is a good point I have not considered before.

@eproxus 

> Would be cool if `equiv` could even be real Erlang code (maybe not possible?):
> 
> ```erlang
> -doc #{equiv => label(T, 0)}.
> label(T) ->
>     label(T, 0).
> ```

I don't think it would be possible given Erlang parser and all kinds of things this would imply - reproducing Erlang code from parsed tree is just plain impossible with the current OTP parser (and I say that as maintainer of one of the formatters).

**Comment by eproxus** [comment]
@michalmuskala So there is a pre-processor step of sorts that convert them to internal URLs?

This means it’s not standard Markdown anymore, it’s “Markdown-but”. I’m not criticizing, just trying to get the facts right.

**Comment by manuel-rubio** [comment]
Maybe we should consider an improvement in the way the parser is performing its work and letting some module attributes be included in the parsed tree.

**Comment by michalmuskala** [comment]
> @michalmuskala So there is a pre-processor step of sorts that convert them to internal URLs?
> 
> This means it’s not standard Markdown anymore, it’s “Markdown-but”. I’m not criticizing, just trying to understand.

No, there's a post-processing step. In particular this is done by the ExDoc formatter when the documentation is converted to HTML - alongside things like syntax highlighting and other things like that, see [ExDoc documentation](https://hexdocs.pm/ex_doc/readme.html#syntax-highlighting) for details of what is supported. But yes, I agree it's an extension of markdown to support those things. That said, it's fairly natural since people already write things using the syntax it supports, it just adds auto-linking (similar to how raw http links are frequently auto-linked, which is not "pure" markdown per se).



> Maybe we should consider an improvement in the way the parser is performing its work and letting some module attributes be included in the parsed tree.

Unfortunately I don't believe this would be enough. Erlang AST is just lossy and does not contain enough information to reproduce. Starting with things like macros, e.g. you'd probably like `label(T, ?DEFAULT)` to be preserved, to things like `[X | []]` and `[X]` or `"\x61"` and `"a"` having exactly the same representation, and many, many more.

**Comment by eproxus** [comment]
@michalmuskala
> No, there's a post-processing step. In particular this is done by the ExDoc formatter when the documentation is converted to HTML - alongside things like syntax highlighting and other things like that, see [ExDoc documentation](https://hexdocs.pm/ex_doc/readme.html#syntax-highlighting) for details of what is supported. But yes, I agree it's an extension of markdown to support those things. That said, it's fairly natural since people already write things using the syntax it supports, it just adds auto-linking (similar to how raw http links are frequently auto-linked, which is not "pure" markdown per se).

So instead of the author being able to link to some content deliberately, pure formatting syntax is overloaded at render time to generate semantic structure that the author has no control over. Not sure I like that.

If I type:

```erl
% Use the option `queue` set to `true` to enable...
```
> 
I don't want the word `queue` become a link to the OTP module.

In my opinion any alternative we discuss here should be possible to use as an future alternative to document the whole of OTP and I'm personally not convinced Markdown plus some automatic heuristics will be enough.

**Comment by josevalim** [comment]
> I don't want the word queue become a link to the OTP module.

It works in Elixir because module names and variables have distinct casing. Erlang would need to adopt another approach. For example, we also support this in Elixir, which could be a more explicit approach for Erlang:

    [the queue module](`queue`)

But perhaps that doesn't make sense either. You could also do:

    [the queue module](m:queue)

Or any equivalent mechanism.

But you could also just do:

    [the queue module](queue.html)

And that would work with what ExDoc generates. But that would couple to the output of build tools.

This is definitely a cons in the Markdown side - not sure if enough to rule it out. The other extension we have is Admonition Text support (but what we handle is not "broken" Markdown, a regular Markdown tool would just show it unstyled).

**Comment by DianaOlympos** [comment]
For references, only Asciidoc has a natural way to do them, and even that is limited as it would probably need some processing or at least some work on making sure we have "nice" anchors that make sense.

So the fact we have to "extend" markdown (or djot) for that to work well is imho not that big of a problem. I ofc could be wrong and misunderstanding how we want to present anchors in Asciidoc.

For Admonition, i think they could be handled without extensions in Djot (or more particularly, with this being an attribute at the block level, the formatter would decide how to present it.)

Basically i think the question come down to if we want these features to be done by the markup we decide to use or if it is done by the formatter that have more knowledge of how to present the information in the final way to present it, being a console, man page, pdf or html.

**Comment by wojtekmach** [comment]
One really nice thing I found in Rust docs is how it allows generating references:

```
[std] is shorthand for [std](https://doc.rust-lang.org/1.64.0/std/index.html)

[`std`] is shorthand for [`std`](https://doc.rust-lang.org/1.64.0/std/index.html)

[new String][std::string::String.new()] is shorthand for [new String](https://doc.rust-lang.org/std/string/struct.String.html#method.new)
```

Tools like ExDoc can auto-link `` `File` `` but in hindsight I think it would be better if it had supported `` [`File`] `` instead. Exactly so there's no ambiguity between module names and atoms: if we wrote `` [`:file]` ``, we explicitly wanted to make a reference to the module.

If Markdown is adopted, I hope `` [`ref`] `` will be supported.

**Comment by wojtekmach** [comment]
And yes, the above is not standard CommonMark, 
but it is so convenient and definitely in the spirit of Markdown.

**Comment by paulo-ferraz-oliveira** [comment]
@starbelly, is there any publicly available document where the pros and cons of the different approaches are being listed, or is this the output of a future WG meeting? I see a very healthy discussion above, and believe the choice will be made based on a list of pros and cons. 😄

**Comment by starbelly** [comment]
> @starbelly, is there any publicly available document where the pros and cons of the different approaches are being listed, or is this the output of a future WG meeting? I see a very healthy discussion above, and believe the choice will be made based on a list of pros and cons. 😄

@paulo-ferraz-oliveira  WIll make one today.

**Comment by starbelly** [comment]
I went silent on this one, but I did think a lot. I'm not going to make a pros and cons list myself, I'm not sure it's worth doing, if you search there's some out there in the wild but they are very slanted with preferences and opinions, even the comparison on asciidoctor is not great as it compares asciidoc with plain ole markdown vs common mark or gfm.

 I think asciidoc is very cool, but it's got a lot of bells and whistles which I think are great, but also maybe YAGNI too unless you're writing a book. That's the history of that project though right? It was intended to be a replacement for docbook, not markdown. One thing that bugs me about asciidoc is that there is no spec. One is being worked on now, but that still isn't done as far as I can tell.  
 
 I can't say much for djot,  so I will not. 
 
Markdown on the other hand : 

-  Is ubiquitous .
- There actually are specs for common mark and gfm, which are quite simple. We would surely arrive at erlang or beam flavored markdown if you will, but that's going to happen with asciidoc anyway.
- There's existing reference implementations in our ecosystem. The lift for an asciidoc implementation seemed to be a heavier one to me, this kind of goes back to all the features of asciidoc, that's nice if you're using ascidoctor, but someone would still have to implement all of that and I'd wager between a markdown lib with extensions for our needs vs an asciidoc lib with extensions for our needs, the markdown lib is going to be simpler. 
- It's hard to argue with success, all you have to do is look at the elixir eco system and ex_doc IMO. When I'm writing docs in Elixir, it's feels effortless and I enjoy doing it.  
-  IMHO a nice thing with the adoption of markdown is having a common documentation format between languages (i.e., erlang and elixir). It wouldn't (couldn't) be exactly the same, but I think it would feel close enough, such that people going back and forth between the two experience less of a context switch in that department.

If we are to submit an EEP I feel there needs to be consensus on this. The original idea was to submit an EEP with options, but I'm not sure in retrospect that makes a lot of sense. It may make more sense to hear further from @KennethL on this subject.

**Comment by lpil** [comment]
> IMHO a nice thing with the adoption of markdown is having a common documentation format between languages (i.e., erlang and elixir).

Gleam, Purerl, and Hamler use markdown here too.

**Comment by DianaOlympos** [comment]
Regarding Djot:

- It is mostly looking like Markdown. The main goal was not to go too far from Markdown
- It is easier to parse. This is the main reason the creator of pandoc invented it.
- It is made to be extended (with attributes) so that people using what has been parsed can extend it with their own formatters and transform it. Not too surprising for someone that handles pandoc. Arbitrary attributes can be added to any block so that follow-up AST transform can use it.

Other than that, it is the same thing as markdown. I could. I would say that if we want something quickly, let's go for Markdown. I think Djot is slightly nicer to extend for everything we want to add "on top" of markdown, but markdown is here and it works.

Otherwise, I am interested in writing an erlang (or elixir) parser for Djot, but it is not the most urgent thing for me. If it was something the EEF want to explore, I am happy to make a proposal to spend time on it. Otherwise, it will have to wait.

**Comment by erszcz** [comment]
Djot seems to be the best candidate out there. I've skimmed over [Djot's docs](https://djot.net/) and the [quickstart for Markdown users](https://github.com/jgm/djot/blob/1e4c19517a79d190eb5f6abe31055d989950a08a/doc/quickstart-for-markdown-users.md) and for those who don't need anything else it's just that - Markdown.

For @peerst it has tables ;)

It's possible to embed any other formats in it, either inline or in block form. Together with attributes, which can be used to add arbitrary information to elements/block, this should allow for bridging the gaps with the other formats. Anything needed for "erlang or beam flavored markdown" could be expressed using these mechanisms and it would still comply with Djot's definition instead of being yet another standard/flavour.

The only missing bit is an implementation in any of the BEAM languages.

**Comment by DianaOlympos** [comment]
Thanks, this corresponds to my own analysis. It is really the best of both worlds imho.

I started working on a parser today, but that will be my first ever parser so don't expect anything impressive fast. It is also in elixir, because quite simply, for my first parser, I went for what was easy. I can always port it later if need be.

I do think it is relatively easy to parse, the parser is made to be linear with no backtracking, so it is relatively easy and avoids a lot of the ambiguities of markdown.

**Comment by KennethL** [comment]
Even if both Djot and AsciiDoc have better features I think Markdown (or some popular flavor of it) is the one to choose. The use of Markdown and the tools supporting it is overwhelming compared with the other two.

Djot is a bit interesting since a subset of it (which is also the natural way of writing) is also accepted as valid Markdown. But I don't think that will work in practice.

Elixir is using Markdown so I don't know why it would be a worse fir for Erlang? 
I don't think Elixir will change to something else than Markdown.

**Comment by starbelly** [comment]
Unless there's disagreement, we can begin working on an EEP. Thoughts?

**Comment by michalmuskala** [comment]
We already have one EEP in this area, that I think is actually pretty solid.  We should probably work on amending it, if changes are needed 

https://github.com/erlang/eep/blob/master/eeps/eep-0059.md

**Comment by josevalim** [comment]
EEP 59 does not discuss format on purpose. I think we should submit a different one for format and perhaps amend EEP 59 to point to the format one? And there should be another one for doctests if EEP 59 is ever to be accepted.

**Comment by starbelly** [comment]
I think another EEP makes sense that can reference EEP 59 and vice versa. If EEP 59 was intentionally left ambiguous then it should remain so IMHO.

**Comment by hauleth** [comment]
Reviving it a little, but I want to add some of points. My main gripe with Markdown is that it was created for writing [blogposts with *minimal* formatting](https://daringfireball.net/projects/markdown/). That mean that Markdown is **very** HTML centric and inflexible to extending it *in backwards compatible way*.

Even ignoring the ambiguity in the parsers and stuff that make MD enormously painful to write sometimes (the amount of spaces that you need to use to continue the bullet point is always pain for me), it is terrible when we want to produce anything other than HTML. Erlang documentation is produced in at least 3 formats (that I am aware of) - terminal, PDF and HTML. Markdown/CommonMark fills only one of this gaps.

About extending the MD, someone uses fenced blocks for that, so you can write:

~~~
```mermaid
graph TD;
    A-->B;
    A-->C;
    B-->D;
    C-->D;
```
~~~

To get

```mermaid
graph TD;
    A-->B;
    A-->C;
    B-->D;
    C-->D;
```

But then is the question - how I can have a block of coloured Mermaid code? Not generated diagram, but the direct code? There is no answer for that, as MD/CM do not allow us to differentiate between these two (also, this again, is not PDF-generation/static-HTML friendly solution, as most implementations require JS code to transform such code blocks to diagrams). It also make it hard to add custom plugins.

MD (and even CommonMark) miss some IMHO substantial features, that are super useful for generating documentation (and as mentioned above, there is no way to add it in backwards compatible ways):

- footnotes
- math
- document metadata (which would be useful if we want to support external documents for metadata)
- tables (this is pretty extensively used in OTP IIRC)
- figure captions
- in-document references
- element metadata (f.e. ExDoc needed to come up with their own syntax for admonition blocks)

Of course, we can use one of the many available extensions to MD/CM to provide that features, but that IMHO brings us back to the beginning, as we end with format that is only partially "familiar", as any feature beyond basics require adding "own sugar" that make it non-backward compatible nor cross-implementation compatible.

I like, and I use, Markdown a lot, but for its original intended purpose, as it works great in that niche. But for documenting anything it is for me super painful solution, as it lacks so many features that I would like to use, but after Nth way to hack around its problems, I just give up, and I use easier for me approach instead of easier for reader approach.

Does ASCIIDoc is painful to write sometimes? Yes, it is, but it is something that I can live with as it has features that make it more suitable for documentation (especially as it is meant to be textual representation of DocBook format).

My preferred format would be probably Djot there, as it is familiar enough to "popular" MD/CM, but is more flexible at the points that are useful for documentation and allows to define custom extensions (that would fit the use case for the EDoc macros, but more flexible).

Obviously we can still support MD/CM or other formats thanks to EEP 59 flexibility, but for "default" format I think that MD/CM is a little bit too inflexible and HTML-centric.

**Comment by essen** [comment]
@hauleth gave a great summary of the problem with Markdown. The reason I went with Asciidoc is precisely because Markdown only cares about HTML. It wasn't particularly difficult to go with Asciidoc and generate various output formats (HTML, PDF, man). Going from Asciidoc to doc-chunks isn't particularly difficult either.

Regardless of what you choose you will need some rules on top to restrict what can be used in the documentation anyway. For example I doubt documentation patches that embed HTML where not necessary would be accepted, something that Markdown makes real easy.

**Comment by KennethL** [comment]
We are choosing Markdown because we think it fulfill our requirements in a decent way and its use is widespread. I can agree that asciidoc is a better choice on pure technical merits but not when it comes to available tools etc.
@essen At least my intention with using Markdown for documentation of OTP is to only allow a subset of it and embedded HTML should be completely forbidden. The idea is still to have the documentation source written once and then having backends producing html, doc-chunks and possibly epub. The writer should not and must not bother about layout only about content.

**Comment by KennethL** [comment]
I should say that the working assumption is to use a subset of Markdown (without ebedded html and possibly other restrictions as well. It is important that the documentation written in OTP Markdown flavour can be handled by generic Markdown tools but I don't think it is important being able to use whatever Markdown text inside the Erlang/OTP documentation.
@hauleth Tables are not used at all in the Erlang/OTP API documentation since it is not supported there. So there is no big backlog of tables to port into the so called reference manuals. In users guides there are tables and I am sure we find a way to express them with Markdown.

**Comment by josevalim** [comment]
I agree with @hauleth concerns but I am not sure if tooling is one of them. There are many tools to generate websites/HTML from Markdown and ePUBs are HTML pages. You can use pandoc or more specific tools to convert markdown to man pages as well as use Pandoc (or Calibre) to convert epub to pdf.

The biggest concern is rather that Markdown may not have a construct to express a given feature from a specific format (such as tables) but I have never seen tooling itself as an issue.

**Comment by essen** [comment]
The problem with Markdown tooling comes from the lack of ability of Markdown to write structured content and so you need some glue to make it all go together. Pandoc only gets you so far. If you only consider the function references, Markdown is probably fine. But for the user guides? Definitely not. For that you need a format where you can structure the content. Asciidoc works well for that, I used it successfully, but I wouldn't be surprised if there was a better alternative today. The doc markup space is still seeing a lot of innovation.

**Comment by garazdawi** [comment]
> The doc markup space is still seeing a lot of innovation.

I think this is really the core of the issue. We want to choose something that has a high chance of still being relevant 10-20 years from now. Last time we choose XML and in hindsight that may not have been the best choice, but it worked well enough for a long time. Whatever it is we choose now, we will not change again in a long time, so the best choice (IMO) is a tradeoff between flexibility, tooling and staying power/popularity.

**Comment by josevalim** [comment]
@essen you can also structure the content outside of markup language by passing the "index" as an Erlang data structure. You can check [Ecto docs](https://hexdocs.pm/ecto/getting-started.html) and the [generated ePUB](https://hexdocs.pm/ecto/Ecto.epub) as examples. So it is definitely possible.

I assume with AsciiDoc I would be able to keep the index in AsciiDoc itself? Although I think [Obsidian](https://obsidian.md/) generates all kinds of structures and documents by traversing the Markdown links. Unsure if it somehow "extends" Markdown for that.

**Comment by essen** [comment]
Of course you can do it outside, but that's the glue I am talking about. With Asciidoc you build a structured document to begin with, similar to what Docbook does. Depending on what you need, you may output a table of content, a list of images or snippets, a glossary and more, automatically (and optionally). You can cross reference content without having to worry about where that content is located in the source files. You can include files. And it's all part of the markup, not some external tool you need to plug into or build yourself.

**Comment by josevalim** [comment]
Thanks @essen, I appreciate the clarification.

**Comment by erszcz** [comment]
I'd like to reiterate the points stated about [djot](https://github.com/jgm/djot), which is mostly Markdown on the surface, but [standardises syntax of elements not present in Markdown](https://htmlpreview.github.io/?https://github.com/jgm/djot/blob/master/doc/syntax.html), such as tables, definition lists, footnotes, references, math formulas, and others present in AsciiDoc. One of the design goals states that:

> There should be generic containers for text, inline content, and block-level content, to which arbitrary attributes can be applied. This allows for extensibility using AST transformations.

djot is authored by John MacFarlane, the author of Pandoc, and comes with reference implementations in Lua and TypeScript. The latter allows to convert between djot and Pandoc internal formats, therefore opening the possibilities to export to other formats (AsciiDoc / Docbook / epub).

It also has [an online playground, where I've played with some hypothetical ad-hoc Erlang-specific attributes](https://djot.net/playground/?text=%7B%23djot-example%7D%0A%23+Djot+example+with+some+ad-hoc+Erlang-specific+syntax%0A%0A%7B%23introduction%7D%0AThis+is+the+introductory+paragraph+taken+from+Djot+readme%2C+with%0Aan+identifier+%60introduction%60.%0A%0A+++++++++++%7B.important+color%3D%22blue%22+%23heading%7D%0A%23%23+Level-2+heading%0A%0AThe+word+*atelier*%7Bweight%3D%22600%22%7D+is+French.%0A%0A%7B%23warning+.sidebar%7D%0A%3A%3A%3A+Warning%0AThis+is+a+warning.%0AHere+is+a+word+in+%5Bfran%C3%A7ais%5D%7Blang%3Dfr%7D.%0A%3A%3A%3A%0A%0AExample+paragraph+mentioning+an+Erlang+function%2C+%5Blists%3Afoldl%2F3%5D%5B%5D%2C+with+no+custom+attributes.%0A%0AAnother+example%2C+this+time+using+a+hypothetical+set+of+Erlang-specific+attributes+-+a+reference%0Ato+documentation+at+a+custom+location.%0AThe+link+to+the+%5Bshell%5D%7B.m+app%3Dstdlib+docroot%3D%22https%3A%2F%2Ferszcz.github.io%22%7D+module%0Adoc+generated+with+%5BExDoc%5D%28https%3A%2F%2Fgithub.com%2Felixir-lang%2Fex_doc%29.%0AIt+could+be+easily+converted+to+just+%5Bshell%5D%28https%3A%2F%2Ferszcz.github.io%2Fstdlib%2Fshell.html%29+by+a+processor.%0AStill+the+info+that+it%27s+a+module+%28%60.m%60%29%2C+not+a+function+%28%60.f%60%29+or+a+callback+%28%60.cb%60%29%2C%0Athat+it+comes+from+%60stdlib%60%2C+and+that+it%27s+hosted+in+a+non-standard+location+is+kept+in+a+structured+format.&mode=ast&sourcepos=false). Such attributes allow to capture the concepts for which Elixir/ExDoc had to invent custom syntax (callback/type links, admonition blocks mentioned by @hauleth).

I've not used this format or its tooling extensively, but in the context of AsciiDoc and Markdown being the strongest contenders, it just seems to provide the best of both worlds.

**Comment by eproxus** [comment]
@erszcz I agree on all points but want to add the disclaimer that the Djot design is not yet finalized, for what it’s worth.

**Comment by essen** [comment]
Worth a read on this topic: https://docs.asciidoctor.org/asciidoc/latest/asciidoc-vs-markdown/

**Comment by garazdawi** [comment]
> They want to all the nice things that they see in elixir docs, which is achievable in markdown in an erlang project now, but not in edoc.

What are these nice things?

> it seems like this has slowed down and for good reasons.

Anyway, things have not slowed down on our end, they have rather picked up. The last couple of weeks we've been looking at what can be done for the Erlang/OTP docs when using ExDoc. As ExDoc today only supports markdown, that is the markup I've started with.

You can see what I've gotten so far here: https://garazdawi.github.io/otp/

I've written a tool that takes EEP48 doc chunks and converts them to markdown and then inserts them into the modules/guides (actually I stole the EEP48 -> markdown converter I wrote for ErlangLS). You can see the end result in this repo: https://github.com/garazdawi/otp/tree/lukas/erl_docgen/markdown.

Some examples:
* [atomics.erl](https://github.com/garazdawi/otp/blob/lukas/erl_docgen/markdown/erts/preloaded/src/atomics.erl) -> [atomics.html](https://garazdawi.github.io/otp/erts/atomics.html)
* [eep48_chapter.md](https://github.com/garazdawi/otp/blob/lukas/erl_docgen/markdown/lib/kernel/doc/src/eep48_chapter.md) -> [eep48_chapter.html](https://garazdawi.github.io/otp/erl_docgen/doc_storage.html)
* [erl.md](https://github.com/garazdawi/otp/blob/lukas/erl_docgen/markdown/erts/doc/src/erl_cmd.md) -> [erl.html](https://garazdawi.github.io/otp/erts/erl_cmd.html)

There are still a bunch of things that needs addressing, but it is a start and we will continue working in this area.

I've bumped into the problems with Markdown being very un-structured and behaving differently depending on the implementation a bunch of times already... as an example you can have a look at `erl` above and see the difference in how github renders it and ExDoc. Most of it comes from having to use somewhat odd features in Markdown to render definition lists, but there are other things as well...

**Comment by wojtekmach** [comment]
@garazdawi this is very exciting. If you run into any issues or we can do improvements on ExDoc side, please feel free to open up issues and/or ping me directly.

**Comment by garazdawi** [comment]
I've had to do a couple of things to get links to work and added support for cross application page references (`` `p:app:guide#anchor` ``) and module anchors (`` `m:mod#anchor` ``). You can see the current changes I've done here: https://github.com/garazdawi/ex_doc/commits/lukas/fix-erlang-doc-support. I'll submit PR(s) for the fixes once I know this is the route that we want to go.

One of the main issues I have right now is figuring out how to coax ExDoc into nicely document https://www.erlang.org/doc/man/erlang.html#system_info-1 (https://garazdawi.github.io/otp/erts/erlang.html#system_info/1) and similar functions.

**Comment by josevalim** [comment]
> One of the main issues I have right now is figuring out how to coax ExDoc into nicely document https://www.erlang.org/doc/man/erlang.html#system_info-1 (https://garazdawi.github.io/otp/erts/erlang.html#system_info/1) and similar functions.

For Elixir, the header in ExDoc is the actual signature `foo(arg1, arg2)`. We should probably do the same for Erlang? And we can allow multi signatures too, exactly as in the Erlang docs. The question is how to customize it. Elixir has a heuristic and allows users to override by doing:

```
@doc "..."
def foo(better_name_than_arg1, options)
```

Erlang does not have something similar. We could perhaps do `-doc #{signature => "...."}`, especially if Erlang will allow multiple `-doc` annotations per function (one per clause)?

**Comment by garazdawi** [comment]
> And we can allow multi signatures too, exactly as in the Erlang docs. 

Do you have an example of this? I've looked around but could not find any.

**Comment by TD5** [comment]
@garazdawi Having looked at your linked examples, I just wanted to chime in an say I am really excited for this! Thanks for taking the time to push this forward

**Comment by josevalim** [comment]
@garazdawi Sorry, what I meant to say is that we _could_ allow, but I don't think we allow it yet. It may require changes to EEP 48. For now, maybe you could go with something like `some_fun(otp_release) | some_fun(another_atom)` and then we figure out the best way to break that into

```
some_fun(otp_release)
some_fun(another_atom)
```

in the future?

**Comment by garazdawi** [comment]
In an ideal world I think I would like to do something like this.

In the erlang code we have this:

```erlang
-doc "Returns a list of tuples with information about miscellaneous allocated memory areas.

Each tuple contains an atom describing the type of memory as first element and the amount of allocated
memory in bytes as second element. When information about allocated and used memory is present, also
a third element is present, containing the amount of used memory in bytes.

`erlang:system_info(allocated_areas)` is intended for debugging, and the content is highly
implementation-dependent. The content of the results therefore changes when needed without prior notice.

Notice that the sum of these values is *not* the total amount of memory allocated by the emulator. Some
values are part of other values, and some memory areas are not part of the result. For information about
the total amount of memory allocated by the emulator, see [`erlang:memory/0`].
...
".
-doc #{ since => "OTP-R12B" }.
-spec erlang:system_info(allocated_areas) -> [ tuple() ].

-doc "Returns `{Allocator, Version, Features, Settings}`, where:

* `Allocator` corresponds to the `malloc()` implementation used. If `Allocator` equals `undefined`,
  the `malloc()` implementation used cannot be identified. `glibc` can be identified.
* `Version` is a list of integers (but not a string) representing the version of the `malloc()`
  implementation used.
* `Features` is a list of atoms representing the allocation features used.
* `Settings` is a list of subsystems, their configurable parameters, and used values. Settings can
  differ between different combinations of platforms, allocators, and allocation features. Memory
  sizes are given in bytes.

See also \"System Flags Effecting erts_alloc\" in `erts_alloc(3)`.
".
-doc #{ since => "OTP-R12B" }.
-spec erlang:system_info(allocator) ->
                 {Allocator, Version, Features, Settings} when
      Allocator :: undefined | glibc,
      Version :: [non_neg_integer()],
      Features :: [atom()],
      Settings :: [{Subsystem :: atom(),
                    [{Parameter :: atom(),
                      Value :: term()}]}].
```

Note that I've extended the `-spec` syntax to allow the same function to be defined using multiple attributes (when I asked the compiler team if this was a good idea, they had a panic attack and ran away....).

The above would generate two entries in the doc chunk, exactly like what we have done today for these chunks from xml. When generating docs the order of the entries will be important as otherwise we cannot match the `-spec` in the AST with the `{function,F,A}` entry in the doc chunk. (We could also include the spec in the metadata, like we've done for `application/erlang+html`).

ExDoc would then take these and generate the docs as normal, the only difference would be that it would be possible to expand the function in the sidebar to select which specific clause you want. Or there is a "summary" generated at the anchor for the function that links to all the individual clauses. Or both.

It may be necessary to add some kind of documentation prototype, i.e. 

```erlang
-doc erlang:system_info(AllocatorInformation).
```

Though I would like to avoid that if possible.

**Comment by josevalim** [comment]
> It may be necessary to add some kind of documentation prototype, i.e.

I think this is the hard part. A spec is unfortunately too verbose to use as the header (we tried it in the past and your example shows that too). So unless you can somehow infer the prototype from the spec or the definition, I am unsure. :(

**Comment by essen** [comment]
Just have, as a convention, the first line of the `-doc` block be the function definition. I think your ideal is messy though. Not only is documentation and code mixed up, but the "since" map is on a separate `-doc` line (makes it easier to mess up when adding more clauses in the documentation). The current way of doing things is very good, I wouldn't move too far away from it. Just replace the XML with another markup...

About the Markdown problems, good luck. You're going to need it.

**Comment by garazdawi** [comment]
> Just have, as a convention, the first line of the `-doc` block be the function definition. 

hmm, yes, that could work.

> I think your ideal is messy though. Not only is documentation and code mixed up, but the "since" map is on a separate -doc line (makes it easier to mess up when adding more clauses in the documentation). The current way of doing things is very good, I wouldn't move too far away from it. Just replace the XML with another markup...

It is indeed messy, though I'm not sure the current way is any better. For example the current [`erlang:system_info/1`](https://www.erlang.org/doc/man/erlang.html#system_info-1) is incorrect because someone added clauses without updating the docs correctly. In my proposal I've tried to address this by placing the docs in relation to where the `-spec` is located. Though I'm sure I've introduced a couple of other weaknesses that I've yet to think about in the process.

**Comment by essen** [comment]
I don't think anything can really be done about doc updates. Ultimately it comes down to being vigilant when working on the relevant code. Even if the doc was the line above the code change, you'd have to remain vigilant to avoid forgetting to update the documentation. Automated testing of examples in the docs help but only get you so far. So I view it as a non-issue.

**Comment by DianaOlympos** [comment]
> Note that I've extended the `-spec` syntax to allow the same function to be defined using multiple attributes (when I asked the compiler team if this was a good idea, they had a panic attack and ran away....).

Yeah uh. I strongly recommend considering `-spec` to be frozen in time and something we should all consider deprecated. If you want something like that, you better invent a new separate way to spec functions, like the elixir "type system" research is doing. Far too much stuff and tech debt is attached to `-spec` being *exactly* the way it is today.

**Comment by hauleth** [comment]
[Interesting article comparing callouts syntax](https://toast.al/posts/softwarecraft/2023-08-29_semantic-markup-for-callouts)

**Comment by garazdawi** [comment]
We have (finally) opened [a PR](https://github.com/erlang/otp/pull/7936) for the EEP-59 changes and with it the compiler now can insert any text into the beam files. Once that is merged, it will not be long before the PR that will change all of our docs appear.

**Comment by starbelly** [comment]
I think it's safe to say we can close this "agenda" item now... and that was an amazing and splendid turn of events.
